File: src/main/java/com/bokeunjeong/portfolio/controller/PortfolioController.java
package com.bokeunjeong.portfolio.controller;

import com.bokeunjeong.portfolio.model.entity.Contact;
import com.bokeunjeong.portfolio.model.entity.Link;
import com.bokeunjeong.portfolio.model.entity.Project;
import com.bokeunjeong.portfolio.model.entity.Skill;
import com.bokeunjeong.portfolio.service.PortfolioService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.ResponseBody;

import java.util.List;
import java.util.Map;

@Slf4j
@Controller
public class PortfolioController {

    private final PortfolioService portfolioService;

    @Autowired
    public PortfolioController(PortfolioService portfolioService) {
        this.portfolioService = portfolioService;
    }


    @GetMapping("/")
    public String index(Model model) {

        Map<String, List<Project>> projects = portfolioService.getProjectsByType();
        Map<String, List<Skill>> workSkills = portfolioService.getSkillsByTypeOf("Work");
        Map<String, List<Skill>> personalSkills = portfolioService.getSkillsByTypeOf("Personal");
        Map<Boolean, List<Link>> linksByColumn = portfolioService.getAllLinks();

        log.info(projects.toString());
        model.addAttribute("introductions", portfolioService.getAllIntroductions());
        model.addAttribute("link1", linksByColumn.getOrDefault(true, null));
        model.addAttribute("link2", linksByColumn.getOrDefault(false, null));
        model.addAttribute("workProjects", projects.get("Work"));
        model.addAttribute("sideProjects", projects.get("Personal"));
        model.addAttribute("workSkills", workSkills);
        model.addAttribute("personalSkills", personalSkills);
        model.addAttribute("contacts", portfolioService.getContacts());

        return "index";
    }

    @GetMapping("/contacts")
    public @ResponseBody
    List<Contact> listContacts() {

        return portfolioService.getContacts();
    }

    @PostMapping("/email")
    public @ResponseBody
    String sendEmail() {

        return "OK";
    }
}

File: src/main/kotlin/com/bokeunjeong/portfolio/controller/PortfolioController.kt
package com.bokeunjeong.portfolio.controller

import com.bokeunjeong.portfolio.dto.PortfolioContactDto
import com.bokeunjeong.portfolio.dto.PortfolioProjectDto
import com.bokeunjeong.portfolio.dto.PortfolioSkillDto
import com.bokeunjeong.portfolio.service.PortfolioService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping


@Controller
class PortfolioController {

    val log = LoggerFactory.getLogger(javaClass)

    @Autowired
    lateinit var portfolioService: PortfolioService

    var colors: Map<String, String> = mapOf(
            Pair("Language", "bg-info"),
            Pair("Framework", "bg-success"),
            Pair("Database", "bg-danger"),
            Pair("Tool", "bg-primary"),
            Pair("Design", "bg-warning")
    )

    @GetMapping
    fun index(model: Model): String {

        var projects: List<PortfolioProjectDto> = portfolioService.findAllProjects()
        model.addAttribute("projects", projects)

        var skills: LinkedHashMap<String, MutableList<PortfolioSkillDto>> = getSkills()
        model.addAttribute("skills", skills.entries)

        var contacts: List<PortfolioContactDto> = portfolioService.findAllContacts()
        model.addAttribute("contacts", contacts)

        return "index"
    }

    private fun getSkills(): LinkedHashMap<String, MutableList<PortfolioSkillDto>> {

        var skills = LinkedHashMap<String, MutableList<PortfolioSkillDto>>()
        for (skill: PortfolioSkillDto in portfolioService.findAllUsingSkills()) {
            skill.color = colors.get(skill.type).toString()
            if (!skills.containsKey(skill.type)) {
                skills.put(skill.type, mutableListOf<PortfolioSkillDto>())
            }
            var list: MutableList<PortfolioSkillDto>? = skills.get(skill.type)
            list?.add(skill)
        }
        log.info(skills.toString())

        return skills
    }
}

